# 10 - Engagement Intelligence

**Objective:** Analyse employee engagement from the performance/engagement dataset.

- Average engagement score
- Engagement broken down by department
- List of lowest-engagement employees for HR review

In [1]:
import pandas as pd
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

DATA_PATH = "../data/raw"
df = pd.read_csv(f"{DATA_PATH}/hr_performance_engagement.csv")
print(f"Shape: {df.shape}")
print(f"Columns: {list(df.columns)}")
df.head()

Shape: (2845, 28)
Columns: ['Employee ID', 'StartDate', 'Title', 'BusinessUnit', 'EmployeeStatus', 'EmployeeType', 'PayZone', 'EmployeeClassificationType', 'DepartmentType', 'Division', 'DOB', 'State', 'GenderCode', 'RaceDesc', 'MaritalDesc', 'Performance Score', 'Current Employee Rating', 'Survey Date', 'Engagement Score', 'Satisfaction Score', 'Work-Life Balance Score', 'Training Date', 'Training Program Name', 'Training Type', 'Training Outcome', 'Training Duration(Days)', 'Training Cost', 'Age']


,Employee ID,StartDate,Title,BusinessUnit,EmployeeStatus,EmployeeType,PayZone,EmployeeClassificationType,DepartmentType,Division,...,Engagement Score,Satisfaction Score,Work-Life Balance Score,Training Date,Training Program Name,Training Type,Training Outcome,Training Duration(Days),Training Cost,Age
0,3427,20-Sep-19,Production Technician I,CCDR,Active,Contract,Zone C,Temporary,Production,Finance & Accounting,...,1,2,3,15-Jul-23,Leadership Development,Internal,Failed,2,606.11,50
1,3428,11-Feb-23,Production Technician I,EW,Active,Contract,Zone A,Part-Time,Production,Aerial,...,2,1,5,12-Sep-22,Customer Service,External,Incomplete,4,673.02,58
2,3429,10-Dec-18,Area Sales Manager,PL,Active,Full-Time,Zone B,Part-Time,Sales,General - Sga,...,1,2,1,13-Aug-22,Leadership Development,External,Failed,2,413.28,27
3,3430,21-Jun-21,Area Sales Manager,CCDR,Active,Contract,Zone A,Full-Time,Sales,Finance & Accounting,...,5,5,4,15-Dec-22,Project Management,External,Completed,3,663.78,23
4,3431,29-Jun-19,Area Sales Manager,TNS,Active,Contract,Zone A,Temporary,Sales,General - Con,...,2,5,3,13-Jul-23,Technical Skills,External,Failed,5,399.03,50


In [2]:
# Find engagement score column
eng_cols = [c for c in df.columns if 'engagement' in c.lower()]
print(f"Engagement columns: {eng_cols}")

# Find department column
dept_cols = [c for c in df.columns if 'department' in c.lower()]
print(f"Department columns: {dept_cols}")

# Find ID column
id_cols = [c for c in df.columns if 'id' in c.lower()]
print(f"ID columns: {id_cols}")

Engagement columns: ['Engagement Score']
Department columns: ['DepartmentType']
ID columns: ['Employee ID']


In [3]:
# Overall engagement statistics
eng_col = eng_cols[0] if eng_cols else None
if eng_col:
    print(f"Engagement Score Column: {eng_col}")
    print(f"  Mean: {df[eng_col].mean():.2f}")
    print(f"  Median: {df[eng_col].median():.2f}")
    print(f"  Std: {df[eng_col].std():.2f}")
    print(f"  Min: {df[eng_col].min()}")
    print(f"  Max: {df[eng_col].max()}")
    print(f"  Count: {len(df)}")

Engagement Score Column: Engagement Score
  Mean: 2.94
  Median: 3.00
  Std: 1.44
  Min: 1
  Max: 5
  Count: 2845


In [4]:
# Engagement by department
dept_col = dept_cols[0] if dept_cols else None
if eng_col and dept_col:
    dept_eng = df.groupby(dept_col)[eng_col].mean().sort_values(ascending=False)
    print("Average Engagement by Department:")
    print(dept_eng.round(2))

Average Engagement by Department:
DepartmentType
Executive Office        3.38
IT/IS                   3.02
Sales                   2.98
Software Engineering    2.96
Admin Offices           2.95
Production              2.91
Name: Engagement Score, dtype: float64


In [5]:
# Lowest engagement employees
id_col = id_cols[0] if id_cols else None
if eng_col:
    cols_to_show = [eng_col]
    if dept_col:
        cols_to_show.append(dept_col)
    if id_col:
        cols_to_show.insert(0, id_col)
    
    lowest = df.nsmallest(10, eng_col)[cols_to_show]
    print("Top 10 Lowest Engagement Employees:")
    print(lowest.to_string(index=False))

Top 10 Lowest Engagement Employees:
 Employee ID  Engagement Score    DepartmentType
        3427                 1 Production       
        3429                 1             Sales
        3433                 1             Sales
        3434                 1             Sales
        3439                 1             Sales
        3440                 1             Sales
        3444                 1             Sales
        3466                 1             IT/IS
        3477                 1             IT/IS
        3483                 1             Sales


## Findings

- Engagement scores are aggregated from multiple survey periods per employee
- The engagement dataset has more rows (2,845) than the attrition dataset (1,470)
- Department-level analysis shows which teams have the lowest engagement
- HR can use the low-engagement list to target retention interventions